public/scripts/logprobs.js

```diff
@@ -0,0 +1,455 @@
+ */
+function withVirtualWhitespace(text, span) {
+    const result = [span];
+    if (text.match(/^\s/)) {
+        result.unshift(document.createTextNode('\u200b'));
+    }
+    if (text.match(/\s$/)) {
+        result.push($(document.createTextNode('\u200b')));
+    }
+    // line breaks are trickier. we don't currently handle consecutive line
+    // breaks or line breaks occuring in between non-whitespace characters, but
+    // tokenizers generally don't produce those anyway.
+
+    // matches leading line break, at least one character, and trailing line break
+    if (text.match(/^\n(?:.|\n)+\n$/)) {
+        result.unshift($('<br>'));
+        result.push($('<br>'));
+    } else if (text.match(/^\n/)) {
+        result.unshift($('<br>'));
+    } else if (text.match(/\n$/)) {
+        result.push($('<br>'));
+    }
+    return result;
+}
+
+/**
+ * saveLogprobs receives an array of TokenLogprobs objects representing the
+ * top logprobs for each token in the active message, and saves it to state.
+ * @param {TokenLogprobs[]} logprobs - message logprob data to save
+ * @param {string | null} continueFrom  - optional prompt text to save as the prefix
+ */
+export function saveLogprobs(logprobs, continueFrom) {
+    convertTokenIdLogprobsToText(logprobs);
+
+    const msgId = chat.length - 1;
+    /** @type {MessageLogprobData} */
+    const data = {
+        created: new Date().getTime(),
+        api: getGeneratingApi(),
+        messageId: msgId,
+        swipeId: chat[msgId].swipe_id,
+        messageLogprobs: logprobs,
+        continueFrom,
+        hash: getMessageHash(chat[msgId]),
+    }
+
+    state.messageLogprobs.set(data.hash, data);
+
+    // Clean up old logprobs data
+    const oldLogprobs = Array.from(state.messageLogprobs.values())
+        .sort((a, b) => b.created - a.created)
+        .slice(MAX_MESSAGE_LOGPROBS);
+    for (const oldData of oldLogprobs) {
+        state.messageLogprobs.delete(oldData.hash);
+    }
+}
+
+function getMessageHash(message) {
+    return getStringHash(JSON.stringify({
+        name: message.name,
+        sid: message.swipe_id,
+        mid: chat.indexOf(message),
+    }));
+}
+
+/**
+ * getActiveMessageLogprobData returns the logprobs data for the active chat
+ * message.
+ * @returns {MessageLogprobData || null}
+ */
+function getActiveMessageLogprobData() {
+    const hash = getMessageHash(chat[chat.length - 1]);
+    return state.messageLogprobs.get(hash) || null;
+}
+
+/**
+ * convertLogprobTokenIdsToText mutates the given logprobs data's topLogprobs
+ * field keyed by token text instead of token ID. This is only necessary for
+ * APIs which only return token IDs in their logprobs data; for others this
+ * function is a no-op.
+ * @param {TokenLogprobs[]} input - logprobs data with numeric token IDs
+ */
+function convertTokenIdLogprobsToText(input) {
+    const api = getGeneratingApi();
+    if (api !== 'novel') {
+        return input;
+    }
```

Comment:
This whole function is a workaround for NovelAI not sending decoded tokens in the logprobs response.  Since we don't have a client-side tokenizer, and we don't want to send HTTP requests to the server-side tokenizer every single time a streaming token comes in, the NAI handler instead builds the `TokenLogprobs` array with numeric tokenIDs, and then we decode them here once streaming is completed.

This required modifying the `decodeTextTokens` backend to support returning an array of token chunks since currently it only gives you back the entire joined string, which doesn't work when we need the individual tokens to show in the logrobs UI.